README.md
# ELS - Enterprise Linux/EuroLinux Sources - libfreehand
 
## Usage:
  Checkout branch or tag. Use `make sources` to recreate  SOURCE0 archive.


src/ChangeLog
2015-06-11  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [ef6f6a4e9ef987859c31b96ae824014215dc5722]

	NEWS

	Change-Id: Ie356430187a403850f87ef4ba5fa6c99b89dbe2e

2015-05-29  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [d12a873411c4dd79dda9e1abe5bea5308ef458a3]

	Apply fake transforms in the right sequence

	Change-Id: I89d19a633e433d80fd940d9ffd948a778457bb76

2015-05-22  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [e5b9c8c030a6fd7001cce07abf089c189745012e]

	Specify the with and height

	Change-Id: I01ca31f7fea20acdcbf6927d659290970db02b80

2015-05-22  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [1385b3a27d951d7be72c0f859f026780d25e74c2]

	Just a version increase so that I don't forget

	Change-Id: Ieb9e9aa95a3b476fcbe7097fd16dcb364bf34eee

2015-05-22  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [2d87343d8e70a38ca07a39ea0ad4c41006792291]

	Weffc++

	Change-Id: Ie88c600871dbb8bc2b728e1412204a240e1b06e7

2015-05-21  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [3f77d59d6debb7b7c23ec5e39bcf21f9c7b11278]

	More debugging of embedded objects

	Change-Id: I04e8ff201e70ff16e9adeda0d24ebefd29612b29

2015-05-21  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [886904f656d3bf592dc959a654341c5f87dff356]

	RVNGBinaryData is a valid type for RVNGPropertyList value

	Change-Id: I5223b3b9e9fdb72802224275f2e25dffed31ac8b

2015-05-21  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [98d7e306332980adef1e85cb8bb36a8a0d3f4331]

	Emulate ContourFill and NewContourFill by radial gradient

	Change-Id: Ia0d82323ba8fbf7c7b2d62a9b294443f741fc15a

2015-05-21  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [b880cb857dbe8b9055b12c2a04553672b65f909c]

	Better theory about ClipGroup

	Change-Id: I6e8fb82a304a71ac94f58a582c43388cd843517c

2015-05-20  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [80b2741b20c5bd80708accf7bb83a4d1ab52fba8]

	PatternFill

	Change-Id: I334186ef500ebc24620afe3868a120dfe05d6940

2015-05-19  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [aee9a111636a64c2eb792e55ef0260473ff8962e]

	Some more fixes for content fills

	Change-Id: I603eeee0c1d6bb0e56d12e9df210250e1e1eaf95

2015-05-19  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [a4fe7c168396cafed40cb47093b6142f25cd501a]

	Detect common file-types so that we don't output bogus mimetype always

	Change-Id: Icf8a814813ad8b4bae9c12bde2593db8a26e877f

2015-05-18  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [96d6908160fadeb26c32b2ab737d4ed745372d46]

	Use the FHBoundingBox struct

	Change-Id: If7713f849985758166dfb445a7e835dacee8bc83

2015-05-18  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [d4bdcb2bcdc4774e1b40363134652c74f4c884f5]

	Initialize the bounding box properly

	Change-Id: I7b5271866d9d4afe75284401be23b59db8f4c58b

2015-05-18  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [9ae62324f4c48caacfb852881672408d25f4b15a]

	Some fixes around TileFill

	Change-Id: I8d0c0aa7f50b5749e9663cb231c6483ebb7b1521

2015-05-18  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [5a6bcb2c4de7d151b89d22a86e9836894882118f]

	Putting some belts and braces around the svg generation

	Change-Id: I1d8ae5ab101759da830472802f0885bc45d3e5b5

2015-05-18  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [c49d059e90d1701a1eb981d3109bc403fd9dc918]

	Another theory

	Change-Id: I5c6499fbc21b29154dff0a2029941e90ff984478

2015-05-18  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [74833537035d4f21a910b13cf01022ecec616fed]

	Read the raw number

	Change-Id: Icce4bf1b1ab4bd44c5c45de0fca9a91a1f015a41

2015-05-18  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [bcbb7e257a9074908f4d8b993ec276df54350e31]

	Don't bother generating SVG from group id == 0

	Change-Id: I7b1503ca8dd19cb82ce756f023353b3c9b73f4a5

2015-05-18  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [b57584f9b546b677a72fa94b506b096d7ec2c195]

	LibreOffice fills only closed paths

	Change-Id: Ic00cb589c28d89475e40f258ed4ade162c3cac6d

2015-05-18  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [c63103583cfa82e17573eb9229571137ad523e32]

	Some TileFill progress

	Change-Id: I776820b0ca572dd512939460069d84deb3f256f1

2015-05-18  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [6aaae771639184f23f175cf886844564f59cdda1]

	RadialFillX

	Change-Id: Ifc2b29522feba6afdd3580180426d91a7d25b98d

2015-05-17  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [b78fa71bb976dcc5521d3f92d69663032b1b4139]

	Output SymbolInstance

	Change-Id: I57a9d6834ade9cd6603a8bf1fd81adf7ef6d9230

2015-05-17  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [1da9f329d0ed20390b0ec854596912024ca7b509]

	Parsing and collecting SymbolClass + SymbolInstance

	Change-Id: I351d4f858a17be67e12fd46a7a42edb8936b03a4

2015-05-17  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [3d633e159ca0f9bd5fd87eabdd87721f29a6911f]

	So the first theory about group transforms was the right one indeed

	Change-Id: If90d8c03c7364f55a0ad9107a074b18cdcbd86de

2015-05-17  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [1470d02436881ab88532f59dc26959699675e148]

	Some half-baked code for TileFill (not working)

	Change-Id: Icb94a0c4e512a2dfcd8304834d966fb96b113455

2015-05-15  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [f22885ca843d860a48fee61135bca7031c2f0976]

	Parse FWGlowFilter and FWShadowFilter and output shadows existing in ODG

	Change-Id: Ie8617f18c94594a97cd54eda732d51fe51870428

2015-05-14  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [a6bf166ea59b72cc7fb2318f75ec04f747a1f513]

	Slightly different logic of composite path to get the even/odd fills right

	Change-Id: I99e9df7241b7e300754d19f9d8f194a69b073dca

2015-05-14  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [794eb2d3f4ee92e354527218bbbba0023b9c4d94]

	Some little fix

	Change-Id: I25f9e6b4cc13de37780f24df042a6694c7b30881

2015-05-14  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [5a76004ae6c50e883d4192662f28f11d758ba193]

	Some initial handling of OpacityFilter

	Change-Id: I32de4137444291f485e61e6cede3aebd09569465

2015-05-14  Fridrich Štrba  <fridrich.strba@bluewin.ch>  [344a8148c2c625cb0c085c3932109f015c267896]


src/src/lib/tokens.h
#ifndef __FHTOKENS_HXX__
#define __FHTOKENS_HXX__

const int FH_AGDFONT = 1;
const int FH_AGDSELECTION = 2;
const int FH_ARROWPATH = 3;
const int FH_ATTRIBUTEHOLDER = 4;
const int FH_BASICFILL = 5;
const int FH_BASICLINE = 6;
const int FH_BENDFILTER = 7;
const int FH_BLOCK = 8;
const int FH_BRUSH = 9;
const int FH_BRUSHLIST = 10;
const int FH_BRUSHSTROKE = 11;
const int FH_BRUSHTIP = 12;
const int FH_CALLIGRAPHICSTROKE = 13;
const int FH_CHARACTERFILL = 14;
const int FH_CLIPGROUP = 15;
const int FH_COLLECTOR = 16;
const int FH_COLOR6 = 17;
const int FH_COMPOSITEPATH = 18;
const int FH_CONEFILL = 19;
const int FH_CONNECTORLINE = 20;
const int FH_CONTENTFILL = 21;
const int FH_CONTOURFILL = 22;
const int FH_CUSTOMPROC = 23;
const int FH_DATA = 24;
const int FH_DATALIST = 25;
const int FH_DATETIME = 26;
const int FH_DISPLAYTEXT = 27;
const int FH_DUETFILTER = 28;
const int FH_ELEMLIST = 29;
const int FH_ELEMPROPLST = 30;
const int FH_ELEMENT = 31;
const int FH_ENVELOPE = 32;
const int FH_EXPANDFILTER = 33;
const int FH_EXTRUSION = 34;
const int FH_FHDOCHEADER = 35;
const int FH_FWBEVELFILTER = 36;
const int FH_FWBLURFILTER = 37;
const int FH_FWFEATHERFILTER = 38;
const int FH_FWGLOWFILTER = 39;
const int FH_FWSHADOWFILTER = 40;
const int FH_FWSHARPENFILTER = 41;
const int FH_FIGURE = 42;
const int FH_FILEDESCRIPTOR = 43;
const int FH_FILTERATTRIBUTEHOLDER = 44;
const int FH_GRADIENTMASKFILTER = 45;
const int FH_GRAPHICSTYLE = 46;
const int FH_GROUP = 47;
const int FH_GUIDES = 48;
const int FH_HALFTONE = 49;
const int FH_IMAGEFILL = 50;
const int FH_IMAGEIMPORT = 51;
const int FH_IMPORT = 52;
const int FH_LAYER = 53;
const int FH_LENSFILL = 54;
const int FH_LINEPAT = 55;
const int FH_LINETABLE = 56;
const int FH_LINEARFILL = 57;
const int FH_LIST = 58;
const int FH_MDICT = 59;
const int FH_MLIST = 60;
const int FH_MNAME = 61;
const int FH_MQUICKDICT = 62;
const int FH_MSTRING = 63;
const int FH_MASTERPAGEDOCMAN = 64;
const int FH_MASTERPAGEELEMENT = 65;
const int FH_MASTERPAGELAYERELEMENT = 66;
const int FH_MASTERPAGELAYERINSTANCE = 67;
const int FH_MASTERPAGESYMBOLCLASS = 68;
const int FH_MASTERPAGESYMBOLINSTANCE = 69;
const int FH_MPOBJECT = 70;
const int FH_MULTIBLEND = 71;
const int FH_MULTICOLORLIST = 72;
const int FH_NEWBLEND = 73;
const int FH_NEWCONTOURFILL = 74;
const int FH_NEWRADIALFILL = 75;
const int FH_OPACITYFILTER = 76;
const int FH_OVAL = 77;
const int FH_PSLINE = 78;
const int FH_PANTONECOLOR = 79;
const int FH_PARAGRAPH = 80;
const int FH_PATH = 81;
const int FH_PATHTEXTLINEINFO = 82;
const int FH_PATTERNFILL = 83;
const int FH_PATTERNLINE = 84;
const int FH_PERSPECTIVEENVELOPE = 85;
const int FH_PERSPECTIVEGRID = 86;
const int FH_POLYGONFIGURE = 87;
const int FH_PROCEDURE = 88;
const int FH_PROCESSCOLOR = 89;
const int FH_PROPLST = 90;
const int FH_RADIALFILL = 91;
const int FH_RADIALFILLX = 92;
const int FH_RAGGEDFILTER = 93;
const int FH_RECTANGLE = 94;
const int FH_SKETCHFILTER = 95;
const int FH_SPOTCOLOR = 96;
const int FH_SPOTCOLOR6 = 97;
const int FH_STYLEPROPLST = 98;
const int FH_SWFIMPORT = 99;
const int FH_SYMBOLCLASS = 100;
const int FH_SYMBOLINSTANCE = 101;
const int FH_SYMBOLLIBRARY = 102;
const int FH_TEFFECT = 103;
const int FH_TFONPATH = 104;
const int FH_TSTRING = 105;
const int FH_TABTABLE = 106;
const int FH_TAPEREDFILL = 107;
const int FH_TAPEREDFILLX = 108;
const int FH_TEXTBLOK = 109;
const int FH_TEXTCOLUMN = 110;
const int FH_TEXTEFFS = 111;
const int FH_TEXTINPATH = 112;
const int FH_TILEFILL = 113;
const int FH_TINTCOLOR = 114;
const int FH_TINTCOLOR6 = 115;
const int FH_TRANSFORMFILTER = 116;
const int FH_USTRING = 117;
const int FH_VDICT = 118;
const int FH_VMPOBJ = 119;
const int FH_XFORM = 120;

const int FH_TOKEN_COUNT = 121;

const int FH_TOKEN_INVALID = -1;

#endif


src/src/lib/tokenhash.h
/* C++ code produced by gperf version 3.0.4 */
/* Command-line: /usr/bin/gperf --compare-strncmp -C -m 20 ../../src/lib/tokens.gperf  */
/* Computed positions: -k'1,4-5,8' */

#if !((' ' == 32) && ('!' == 33) && ('"' == 34) && ('#' == 35) \
      && ('%' == 37) && ('&' == 38) && ('\'' == 39) && ('(' == 40) \
      && (')' == 41) && ('*' == 42) && ('+' == 43) && (',' == 44) \
      && ('-' == 45) && ('.' == 46) && ('/' == 47) && ('0' == 48) \
      && ('1' == 49) && ('2' == 50) && ('3' == 51) && ('4' == 52) \
      && ('5' == 53) && ('6' == 54) && ('7' == 55) && ('8' == 56) \
      && ('9' == 57) && (':' == 58) && (';' == 59) && ('<' == 60) \
      && ('=' == 61) && ('>' == 62) && ('?' == 63) && ('A' == 65) \
      && ('B' == 66) && ('C' == 67) && ('D' == 68) && ('E' == 69) \
      && ('F' == 70) && ('G' == 71) && ('H' == 72) && ('I' == 73) \
      && ('J' == 74) && ('K' == 75) && ('L' == 76) && ('M' == 77) \
      && ('N' == 78) && ('O' == 79) && ('P' == 80) && ('Q' == 81) \
      && ('R' == 82) && ('S' == 83) && ('T' == 84) && ('U' == 85) \
      && ('V' == 86) && ('W' == 87) && ('X' == 88) && ('Y' == 89) \
      && ('Z' == 90) && ('[' == 91) && ('\\' == 92) && (']' == 93) \
      && ('^' == 94) && ('_' == 95) && ('a' == 97) && ('b' == 98) \
      && ('c' == 99) && ('d' == 100) && ('e' == 101) && ('f' == 102) \
      && ('g' == 103) && ('h' == 104) && ('i' == 105) && ('j' == 106) \
      && ('k' == 107) && ('l' == 108) && ('m' == 109) && ('n' == 110) \
      && ('o' == 111) && ('p' == 112) && ('q' == 113) && ('r' == 114) \
      && ('s' == 115) && ('t' == 116) && ('u' == 117) && ('v' == 118) \
      && ('w' == 119) && ('x' == 120) && ('y' == 121) && ('z' == 122) \
      && ('{' == 123) && ('|' == 124) && ('}' == 125) && ('~' == 126))
/* The character set is not based on ISO-646.  */
#error "gperf generated tables don't work with this execution character set. Please report a bug to <bug-gnu-gperf@gnu.org>."
#endif

#line 5 "../../src/lib/tokens.gperf"
struct fhtoken
{
  const char *name;
  int tokenId;
};

#define TOTAL_KEYWORDS 120
#define MIN_WORD_LENGTH 4
#define MAX_WORD_LENGTH 24
#define MIN_HASH_VALUE 8
#define MAX_HASH_VALUE 161
/* maximum key range = 154, duplicates = 0 */

class Perfect_Hash
{
private:
  static inline unsigned int hash (const char *str, unsigned int len);
public:
  static const struct fhtoken *in_word_set (const char *str, unsigned int len);
};

inline unsigned int
Perfect_Hash::hash (register const char *str, register unsigned int len)
{
  static const unsigned char asso_values[] =
    {
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162,  68,   3,   0,  37,  26,
        2,  25,   2,  42, 162, 162,  37,   0,  27,  69,
        7, 162,  73,  37,   2,   1,  55, 162,  43, 162,
      162, 162, 162, 162, 162, 162, 162,   5,  17,  42,
       67,   2,   0,  32,  10,   1,  82,  20,  28,  61,
       44,  26,  11, 162,   0,  43,   0,  69,   8,  25,
      162,  52, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162, 162, 162, 162, 162,
      162, 162, 162, 162, 162, 162
    };
  register int hval = len;

  switch (hval)
    {
      default:
        hval += asso_values[(unsigned char)str[7]];
      /*FALLTHROUGH*/
      case 7:
      case 6:
      case 5:
        hval += asso_values[(unsigned char)str[4]];
      /*FALLTHROUGH*/
      case 4:
        hval += asso_values[(unsigned char)str[3]];
      /*FALLTHROUGH*/
      case 3:
      case 2:
      case 1:
        hval += asso_values[(unsigned char)str[0]];
        break;
    }
  return hval;
}

static const struct fhtoken wordlist[] =
  {
    {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0},
    {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0},
#line 73 "../../src/lib/tokens.gperf"
    {"MString",FH_MSTRING},
#line 127 "../../src/lib/tokens.gperf"
    {"UString",FH_USTRING},
#line 115 "../../src/lib/tokens.gperf"
    {"TString",FH_TSTRING},
#line 113 "../../src/lib/tokens.gperf"
    {"TEffect",FH_TEFFECT},
#line 59 "../../src/lib/tokens.gperf"
    {"Halftone",FH_HALFTONE},
#line 81 "../../src/lib/tokens.gperf"
    {"MultiBlend",FH_MULTIBLEND},
    {(char*)0, 0},
#line 31 "../../src/lib/tokens.gperf"
    {"ContentFill",FH_CONTENTFILL},
    {(char*)0, 0},
#line 117 "../../src/lib/tokens.gperf"
    {"TaperedFill",FH_TAPEREDFILL},
#line 118 "../../src/lib/tokens.gperf"
    {"TaperedFillX",FH_TAPEREDFILLX},
#line 116 "../../src/lib/tokens.gperf"
    {"TabTable",FH_TABTABLE},
#line 24 "../../src/lib/tokens.gperf"
    {"CharacterFill",FH_CHARACTERFILL},
#line 91 "../../src/lib/tokens.gperf"
    {"Path",FH_PATH},
#line 93 "../../src/lib/tokens.gperf"
    {"PatternFill",FH_PATTERNFILL},
#line 74 "../../src/lib/tokens.gperf"
    {"MasterPageDocMan",FH_MASTERPAGEDOCMAN},
#line 75 "../../src/lib/tokens.gperf"
    {"MasterPageElement",FH_MASTERPAGEELEMENT},
#line 54 "../../src/lib/tokens.gperf"
    {"FilterAttributeHolder",FH_FILTERATTRIBUTEHOLDER},
#line 48 "../../src/lib/tokens.gperf"
    {"FWFeatherFilter",FH_FWFEATHERFILTER},
#line 46 "../../src/lib/tokens.gperf"
    {"FWBevelFilter",FH_FWBEVELFILTER},
#line 78 "../../src/lib/tokens.gperf"
    {"MasterPageSymbolClass",FH_MASTERPAGESYMBOLCLASS},
#line 76 "../../src/lib/tokens.gperf"
    {"MasterPageLayerElement",FH_MASTERPAGELAYERELEMENT},
#line 77 "../../src/lib/tokens.gperf"
    {"MasterPageLayerInstance",FH_MASTERPAGELAYERINSTANCE},
#line 79 "../../src/lib/tokens.gperf"
    {"MasterPageSymbolInstance",FH_MASTERPAGESYMBOLINSTANCE},
#line 27 "../../src/lib/tokens.gperf"
    {"Color6",FH_COLOR6},
#line 119 "../../src/lib/tokens.gperf"
    {"TextBlok",FH_TEXTBLOK},
#line 51 "../../src/lib/tokens.gperf"
    {"FWSharpenFilter",FH_FWSHARPENFILTER},
#line 92 "../../src/lib/tokens.gperf"
    {"PathTextLineInfo",FH_PATHTEXTLINEINFO},
#line 33 "../../src/lib/tokens.gperf"
    {"CustomProc",FH_CUSTOMPROC},
#line 124 "../../src/lib/tokens.gperf"
    {"TintColor",FH_TINTCOLOR},
#line 125 "../../src/lib/tokens.gperf"
    {"TintColor6",FH_TINTCOLOR6},
#line 32 "../../src/lib/tokens.gperf"
    {"ContourFill",FH_CONTOURFILL},
#line 29 "../../src/lib/tokens.gperf"
    {"ConeFill",FH_CONEFILL},
#line 68 "../../src/lib/tokens.gperf"
    {"List",FH_LIST},
#line 123 "../../src/lib/tokens.gperf"
    {"TileFill",FH_TILEFILL},
#line 82 "../../src/lib/tokens.gperf"
    {"MultiColorList",FH_MULTICOLORLIST},
#line 63 "../../src/lib/tokens.gperf"
    {"Layer",FH_LAYER},
#line 89 "../../src/lib/tokens.gperf"
    {"PantoneColor",FH_PANTONECOLOR},
#line 34 "../../src/lib/tokens.gperf"
    {"Data",FH_DATA},
#line 69 "../../src/lib/tokens.gperf"
    {"MDict",FH_MDICT},
#line 70 "../../src/lib/tokens.gperf"
    {"MList",FH_MLIST},
#line 38 "../../src/lib/tokens.gperf"
    {"DuetFilter",FH_DUETFILTER},
#line 28 "../../src/lib/tokens.gperf"
    {"CompositePath",FH_COMPOSITEPATH},
#line 36 "../../src/lib/tokens.gperf"
    {"DateTime",FH_DATETIME},
#line 23 "../../src/lib/tokens.gperf"
    {"CalligraphicStroke",FH_CALLIGRAPHICSTROKE},
#line 65 "../../src/lib/tokens.gperf"
    {"LinePat",FH_LINEPAT},
#line 72 "../../src/lib/tokens.gperf"
    {"MQuickDict",FH_MQUICKDICT},
#line 67 "../../src/lib/tokens.gperf"
    {"LinearFill",FH_LINEARFILL},
#line 50 "../../src/lib/tokens.gperf"
    {"FWShadowFilter",FH_FWSHADOWFILTER},
#line 94 "../../src/lib/tokens.gperf"
    {"PatternLine",FH_PATTERNLINE},
#line 88 "../../src/lib/tokens.gperf"
    {"PSLine",FH_PSLINE},
#line 122 "../../src/lib/tokens.gperf"
    {"TextInPath",FH_TEXTINPATH},
#line 98 "../../src/lib/tokens.gperf"
    {"Procedure",FH_PROCEDURE},
#line 19 "../../src/lib/tokens.gperf"
    {"Brush",FH_BRUSH},
#line 100 "../../src/lib/tokens.gperf"
    {"PropLst",FH_PROPLST},
#line 99 "../../src/lib/tokens.gperf"
    {"ProcessColor",FH_PROCESSCOLOR},
#line 90 "../../src/lib/tokens.gperf"
    {"Paragraph",FH_PARAGRAPH},
#line 26 "../../src/lib/tokens.gperf"
    {"Collector",FH_COLLECTOR},
#line 42 "../../src/lib/tokens.gperf"
    {"Envelope",FH_ENVELOPE},
#line 21 "../../src/lib/tokens.gperf"
    {"BrushStroke",FH_BRUSHSTROKE},
#line 71 "../../src/lib/tokens.gperf"
    {"MName",FH_MNAME},
#line 49 "../../src/lib/tokens.gperf"
    {"FWGlowFilter",FH_FWGLOWFILTER},
#line 18 "../../src/lib/tokens.gperf"
    {"Block",FH_BLOCK},
#line 114 "../../src/lib/tokens.gperf"
    {"TFOnPath",FH_TFONPATH},
#line 106 "../../src/lib/tokens.gperf"
    {"SpotColor",FH_SPOTCOLOR},
#line 107 "../../src/lib/tokens.gperf"
    {"SpotColor6",FH_SPOTCOLOR6},
#line 62 "../../src/lib/tokens.gperf"
    {"Import",FH_IMPORT},
#line 22 "../../src/lib/tokens.gperf"
    {"BrushTip",FH_BRUSHTIP},
#line 96 "../../src/lib/tokens.gperf"
    {"PerspectiveGrid",FH_PERSPECTIVEGRID},
#line 52 "../../src/lib/tokens.gperf"
    {"Figure",FH_FIGURE},
#line 66 "../../src/lib/tokens.gperf"
    {"LineTable",FH_LINETABLE},
#line 121 "../../src/lib/tokens.gperf"
    {"TextEffs",FH_TEXTEFFS},
#line 95 "../../src/lib/tokens.gperf"
    {"PerspectiveEnvelope",FH_PERSPECTIVEENVELOPE},
#line 120 "../../src/lib/tokens.gperf"
    {"TextColumn",FH_TEXTCOLUMN},
#line 17 "../../src/lib/tokens.gperf"
    {"BendFilter",FH_BENDFILTER},
#line 15 "../../src/lib/tokens.gperf"
    {"BasicFill",FH_BASICFILL},
#line 14 "../../src/lib/tokens.gperf"
    {"AttributeHolder",FH_ATTRIBUTEHOLDER},
#line 30 "../../src/lib/tokens.gperf"
    {"ConnectorLine",FH_CONNECTORLINE},
#line 45 "../../src/lib/tokens.gperf"
    {"FHDocHeader",FH_FHDOCHEADER},
#line 35 "../../src/lib/tokens.gperf"
    {"DataList",FH_DATALIST},
#line 43 "../../src/lib/tokens.gperf"
    {"ExpandFilter",FH_EXPANDFILTER},
#line 37 "../../src/lib/tokens.gperf"
    {"DisplayText",FH_DISPLAYTEXT},
#line 101 "../../src/lib/tokens.gperf"
    {"RadialFill",FH_RADIALFILL},
#line 102 "../../src/lib/tokens.gperf"
    {"RadialFillX",FH_RADIALFILLX},
#line 105 "../../src/lib/tokens.gperf"
    {"SketchFilter",FH_SKETCHFILTER},
#line 84 "../../src/lib/tokens.gperf"
    {"NewContourFill",FH_NEWCONTOURFILL},
#line 112 "../../src/lib/tokens.gperf"
    {"SymbolLibrary",FH_SYMBOLLIBRARY},
#line 56 "../../src/lib/tokens.gperf"
    {"GraphicStyle",FH_GRAPHICSTYLE},
#line 41 "../../src/lib/tokens.gperf"
    {"Element",FH_ELEMENT},
#line 53 "../../src/lib/tokens.gperf"
    {"FileDescriptor",FH_FILEDESCRIPTOR},
#line 61 "../../src/lib/tokens.gperf"
    {"ImageImport",FH_IMAGEIMPORT},
#line 16 "../../src/lib/tokens.gperf"
    {"BasicLine",FH_BASICLINE},
#line 58 "../../src/lib/tokens.gperf"
    {"Guides",FH_GUIDES},
#line 87 "../../src/lib/tokens.gperf"
    {"Oval",FH_OVAL},
#line 128 "../../src/lib/tokens.gperf"
    {"VDict",FH_VDICT},
#line 11 "../../src/lib/tokens.gperf"
    {"AGDFont",FH_AGDFONT},
#line 126 "../../src/lib/tokens.gperf"
    {"TransformFilter",FH_TRANSFORMFILTER},
#line 108 "../../src/lib/tokens.gperf"
    {"StylePropLst",FH_STYLEPROPLST},
#line 97 "../../src/lib/tokens.gperf"
    {"PolygonFigure",FH_POLYGONFIGURE},
#line 80 "../../src/lib/tokens.gperf"
    {"MpObject",FH_MPOBJECT},
#line 20 "../../src/lib/tokens.gperf"
    {"BrushList",FH_BRUSHLIST},
#line 130 "../../src/lib/tokens.gperf"
    {"Xform",FH_XFORM},
#line 57 "../../src/lib/tokens.gperf"
    {"Group",FH_GROUP},
#line 55 "../../src/lib/tokens.gperf"
    {"GradientMaskFilter",FH_GRADIENTMASKFILTER},
#line 47 "../../src/lib/tokens.gperf"
    {"FWBlurFilter",FH_FWBLURFILTER},
#line 60 "../../src/lib/tokens.gperf"
    {"ImageFill",FH_IMAGEFILL},
#line 25 "../../src/lib/tokens.gperf"
    {"ClipGroup",FH_CLIPGROUP},
#line 104 "../../src/lib/tokens.gperf"
    {"Rectangle",FH_RECTANGLE},
#line 40 "../../src/lib/tokens.gperf"
    {"ElemPropLst",FH_ELEMPROPLST},
    {(char*)0, 0},
#line 64 "../../src/lib/tokens.gperf"
    {"LensFill",FH_LENSFILL},
#line 110 "../../src/lib/tokens.gperf"
    {"SymbolClass",FH_SYMBOLCLASS},
#line 103 "../../src/lib/tokens.gperf"
    {"RaggedFilter",FH_RAGGEDFILTER},
    {(char*)0, 0}, {(char*)0, 0},
#line 85 "../../src/lib/tokens.gperf"
    {"NewRadialFill",FH_NEWRADIALFILL},
    {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0},
#line 86 "../../src/lib/tokens.gperf"
    {"OpacityFilter",FH_OPACITYFILTER},
#line 13 "../../src/lib/tokens.gperf"
    {"ArrowPath",FH_ARROWPATH},
    {(char*)0, 0},
#line 44 "../../src/lib/tokens.gperf"
    {"Extrusion",FH_EXTRUSION},
    {(char*)0, 0},
#line 39 "../../src/lib/tokens.gperf"
    {"ElemList",FH_ELEMLIST},
#line 83 "../../src/lib/tokens.gperf"
    {"NewBlend",FH_NEWBLEND},
    {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0},
#line 111 "../../src/lib/tokens.gperf"
    {"SymbolInstance",FH_SYMBOLINSTANCE},
    {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0},
    {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0},
#line 129 "../../src/lib/tokens.gperf"
    {"VMpObj",FH_VMPOBJ},
    {(char*)0, 0},
#line 109 "../../src/lib/tokens.gperf"
    {"SwfImport",FH_SWFIMPORT},
    {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0},
    {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0},
    {(char*)0, 0}, {(char*)0, 0}, {(char*)0, 0},
#line 12 "../../src/lib/tokens.gperf"
    {"AGDSelection",FH_AGDSELECTION}
  };

const struct fhtoken *
Perfect_Hash::in_word_set (register const char *str, register unsigned int len)
{
  if (len <= MAX_WORD_LENGTH && len >= MIN_WORD_LENGTH)
    {
      register int key = hash (str, len);

      if (key <= MAX_HASH_VALUE && key >= 0)
        {
          register const char *s = wordlist[key].name;

          if (s && *str == *s && !strncmp (str + 1, s + 1, len - 1) && s[len] == '\0')
            return &wordlist[key];
        }
    }
  return 0;
}
#line 131 "../../src/lib/tokens.gperf"

